Bitcoin etl

samparsky opened the pull request:
Fixes issue #1 
@medvedev1088 Ready for review

medvedev1088 commented on README.md:
```diff
@@ -1,2 +1,168 @@
+## Schema
+
+### blocks.csv
+
```
Could you describe the schema here?

medvedev1088 commented on README.md:
```diff
@@ -1,2 +1,168 @@
+
+
+### transactions.csv
+
```
Could you describe the schema here?

medvedev1088 commented on README.md:
````diff
@@ -1,2 +1,168 @@
+
+1. Run a container out of the image
+    ```bash
+    > docker run -v $HOME/output:/bitcoin-etl/output bitcoin-etl:latest export_all -s 0 -e 5499999 -b 100000 -p https://mainnet.infura.io
````
Should it be export_blocks_and_transactions instead of export_all?

medvedev1088 commented on README.md:
````diff
@@ -1,2 +1,168 @@
+```bash
+> bitcoinetl export_blocks_and_transactions --start-block 0 --end-block 500000 \
+--rpc-pass 'test' --rpc-host 'localhost' --rpc-user 'test' \
+ --blocks-output blocks.csv --transactions-output transactions.csv
````
JSON instead of CSV?

medvedev1088 commented on README.md:
````diff
@@ -1,2 +1,168 @@
+> bitcoinetl get_block_range_for_timestamps --rpc-host= 'localhost' --rpc-user='' --rpc-pass='' --start-timestamp=1325376000 --end-timestamp=1325377000
+```
+
+#### get_block_range_for_timestamps
````
get_block_range_for_date?

medvedev1088 commented on bitcoinetl/providers/auto.py:
```diff
@@ -0,0 +1,30 @@
+
+from bitcoinetl.providers.rpc import BatchRPCProvider
+
+DEFAULT_IPC_TIMEOUT = 60
```
Is it used?

medvedev1088 commented on bitcoinetl/mappers/transaction_mapper.py:
```diff
@@ -0,0 +1,61 @@
+        transaction.confirmations = json_dict.get('confirmations')
+        transaction.time = json_dict.get('time')
+        transaction.blocktime = json_dict.get('blocktime')
+        transaction.vout = json_dict.get('vout')
```
This should be the list of BtcTransactionOutput entities. You need a corresponding mapper.

medvedev1088 commented on bitcoinetl/mappers/transaction_mapper.py:
```diff
@@ -0,0 +1,61 @@
+        transaction.time = json_dict.get('time')
+        transaction.blocktime = json_dict.get('blocktime')
+        transaction.vout = json_dict.get('vout')
+        transaction.vin = json_dict.get('vin')
```
This should be the list of BtcTransactionInput entities. You need a corresponding mapper.

samparsky commented:
@medvedev1088 I have fixed the issues and made the recommended changes

samparsky commented:
@medvedev1088 I have been able to add value to the transaction inputs & made the suggested changes

medvedev1088 commented on bitcoinetl/cli/export_blocks_and_transactions.py:
```diff
@@ -0,0 +1,63 @@
+    if blocks_output is None and transactions_output is None:
+        raise ValueError('Either --blocks-output or --transactions-output options must be provided')
+    
+    if rpc_user is None or rpc_pass is None:
```
This is not needed as the parameter is set required in @click.option()

medvedev1088 commented on README.md:
```diff
@@ -1,2 +1,256 @@
+
+Column                  | Type                  |
+------------------------|-----------------------|
+value                   | string                | 
```
float?

medvedev1088 commented on bitcoinetl/jobs/export_blocks_job.py:
```diff
@@ -0,0 +1,119 @@
+        )
+
+    def add_input_values(self, block):
+        for txIndex, tx in enumerate(block['tx']):
```
txIndex -> tx_index

medvedev1088 commented on bitcoinetl/jobs/export_blocks_job.py:
```diff
@@ -0,0 +1,119 @@
+            txids = list(map(lambda vin: vin["txid"] if "txid" in vin else "", tx['vin']))
+
+            transaction_detail_rpc = list(generate_get_transaction_by_id_json_rpc(txids))
+            transaction_detail_response = self.rpc_provider.make_request(transaction_detail_rpc)
```
I'm not sure JSON RPC batch responses are guaranteed to be returned in the same order as requests in the batch (seems below you rely on it). From the spec examples it looks like the order can be arbitrary and the "id" field needs to be used to match up requests and responses https://www.jsonrpc.org/specification#batch.

medvedev1088 replied:
I'm not sure just adding ids to the request will guarantee the responses will come back in the same order. In the JSON RPC example here https://www.jsonrpc.org/specification#batch the requests come with the ids in a certain order, but the responses are returned in an arbitrary order:

```--> [
        {"jsonrpc": "2.0", "method": "sum", "params": [1,2,4], "id": "1"},
        {"jsonrpc": "2.0", "method": "notify_hello", "params": [7]},
        {"jsonrpc": "2.0", "method": "subtract", "params": [42,23], "id": "2"},
        {"foo": "boo"},
        {"jsonrpc": "2.0", "method": "foo.get", "params": {"name": "myself"}, "id": "5"},
        {"jsonrpc": "2.0", "method": "get_data", "id": "9"} 
    ]
<-- [
        {"jsonrpc": "2.0", "result": 7, "id": "1"},
        {"jsonrpc": "2.0", "result": 19, "id": "2"},
        {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request"}, "id": null},
        {"jsonrpc": "2.0", "error": {"code": -32601, "message": "Method not found"}, "id": "5"},
        {"jsonrpc": "2.0", "result": ["hello", 5], "id": "9"}
    ]
```

In `pythonbitcoinrpc` they just put the incremented ids to requests and then never use them (e.g. to order the responses). Even worse there is no way for us to provide custom ids and handle it ourselves.

A simple way to handle it is to sort the responses by id. I will also submit an issue to pythonbitcoinrpc.

samparsky replied:
@medvedev1088 Yeah true. I would just include include the `authproxy.py` file directly in the codebase and make the change to sort the response by id. Would remove the pythonbitcoinrpc dependency.

samparsky commented:
@medvedev1088 Nice suggestions, I have implemented them

samparsky commented:
@medvedev1088  I have added the `authproxy.py` file & removed the `pythonbitcoinrpc` dependency. Added sorting to the response parsing
https://github.com/blockchain-etl/bitcoin-etl/blob/blockchain-etl/bitcoinetl/providers/authproxy.py#L170

samparsky commented:
@medvedev1088 The PR is okay to be merged now right?

medvedev1088 commented:
@samparsky sorry I've been busy recently. I'll try to review and merge it by the end of this week.
